content/blog/2024-01-11-xilem-backend-roadmap.md

```diff
@@ -0,0 +1,157 @@
+
+Xilem's approach of "you own your children" is a little bespoke. It means the framework has a hard time iterating over the entire widget tree (say, as part of a DOM inspector) unless each Widget specifically implements the iterating method. And it means if you want to target an event at a specific widget (say, the one with keyboard focus), you need to go through the entire ownership chain, using some heuristics to avoid visiting too many widgets.
+
+I believe Widgets should be owned by the library. If your container has children, then the only thing the container will actually own is keys into a structure (probably a slotmap) where the widget is stored. This makes a lot of things easier, like serialization and debugging, but it has an impact on the entire backend. It's an infrastructure investment.
+
+Lately, I've seen more and more dicussion of implementing GUI through an ECS. A lot of that discussion comes from Bevy, which is natural, since the bevy community ~~is filled with ruthless cultists striving to feed ever more sacrifices to the ECS god~~
+(Note: will need to run this joke by the Bevy community)
+TODO
+is intimately familiar with the ECS pattern and putting a lot more importance on UI
+But I've seen discussions about it in the Linebender community too.
+
+Whether we actually want to use ECS is something we still need to research.
+
+The way I understand it, the big draw of ECS is the "S" part, Systems: it's about handling lots of similar data at once in arbitrary order. Which is why eg archetypes in bevy are useful: you're grouping together entities with the same components, so that you can say "execute this code for all entities with component X and component Y" and have this code be executed on arrays of values with minimal indirection or branching. And the "arbitrary order" part makes them work really well with slotmaps: since you can just read from the beginning of the slotmap, you don't pay the double-indirection penalty.
+
+In other words, the ideal use case for ECS is code of the type `for (entity in entities) entity.position += entity.speed`.
+
+GUI is pretty far from that ideal use-case: updates are sparse and should only run on a small number of changed widgets. Order often matters. And I'm not sure archetypes would work in a framework with a large variety of widgets and different attributes.
+
+I think there are two things you really want from a Rust ECS library for GUI: slotmaps, and efficient ways to add and remove components from an entity.
+
+Implementing those is going to be a major undertaking, but one I expect to pay many times over.
```

Comment:
Overall this section is pretty vague and weak. The challenge is that we have a structure (in Druid, Masonry, and Xilem) that is workable. There are challenges to be sure, but we've put thought into how to solve those challenges.

Again, the case hasn't really been made *why* ECS is better or why it will work particularly well (especially as it had been tried in the first iteration of Druid and didn't work out - see https://www.youtube.com/watch?v=4YTfxresvS8 for a talk going into more detail on that).

Finally, if we're doing ECS, why not just adopt bevy-ui? This is a serious question, as Bevy has lots of momentum, and a Xilem-like reactive layer is being built for it.

Reply:
Actually, I never found any analysis of *why* ECS didn't work out for Druid, only that you eventually moved on.

Re: bevy-ui: that's fair, I haven't looked too deeply into it. I suspect it would come with too much of the Bevy ecosystem attached.

Reply:
Continued in this zulip conversation: https://xi.zulipchat.com/#narrow/stream/354396-xilem/topic/Do.20we.20want.20to.20use.20ECS.20in.20Xilem.3F